javaslang/src/main/java/javaslang/collection/RedBlackTree.java

```diff
@@ -412,276 +466,538 @@ public boolean contains(T value) {
+                    final Node<T> tD = new ValueNode<>(node.color(), node.blackHeight(), deletedNode, node.value(), node.right(), node.emptyInstance());
+                    return Tuple.of(tD, false, deletedValue);
                 }
             }
         }
 
-        static <T> Node<T> insert(RedBlackTree<T> tree, T value) {
-            if (tree.isEmpty()) {
-                final Empty<T> empty = (Empty<T>) tree;
-                return new Node<>(RED, 1, empty, value, empty, empty);
+        private static <T> Node<T> insert(Node<T> node, T value) {
+            if (node.isEmpty()) {
+                final Empty<T> empty = (Empty<T>) node;
+                return new ValueNode<>(RED, 1, empty, value, empty, empty);
             } else {
-                final Node<T> node = (Node<T>) tree;
-                final int comparison = node.comparator().compare(value, node.value);
+                final int comparison = node.comparator().compare(value, node.value());
                 if (comparison < 0) {
-                    final Node<T> newLeft = insert(node.left, value);
-                    return (newLeft == node.left)
+                    final Node<T> newLeft = insert(node.left(), value);
+                    return (newLeft == node.left())
                             ? node
-                            : Node.balanceLeft(node.color, node.blackHeight, newLeft, node.value, node.right,
-                            node.empty);
+                            : ValueNode.balanceLeft(node.color(), node.blackHeight(), newLeft, node.value(), node.right(), node.emptyInstance());
                 } else if (comparison > 0) {
-                    final Node<T> newRight = insert(node.right, value);
-                    return (newRight == node.right)
+                    final Node<T> newRight = insert(node.right(), value);
+                    return (newRight == node.right())
                             ? node
-                            : Node.balanceRight(node.color, node.blackHeight, node.left, node.value, newRight,
-                            node.empty);
+                            : ValueNode.balanceRight(node.color(), node.blackHeight(), node.left(), node.value(), newRight, node.emptyInstance());
                 } else {
                     // DEV-NOTE: Even if there is no _comparison_ difference, the object may not be _equal_.
                     //           To save an equals() call, which may be expensive, we return a new instance.
-                    return new Node<>(node.color, node.blackHeight, node.left, value, node.right, node.empty);
+                    return new ValueNode<>(node.color(), node.blackHeight(), node.left(), value, node.right(), node.emptyInstance());
                 }
             }
         }
 
         private static boolean isRed(RedBlackTree<?> tree) {
-            return !tree.isEmpty() && ((Node<?>) tree).color == RED;
+            return !tree.isEmpty() && ((ValueNode<?>) tree).color == RED;
```

Comment:
I see no benefit from adding the Node interface:

```java
interface RedBlackTree
    ^
    |
interface Node
    ^
   / \
Empty ValueNode
```

There are still type casts in the code (see line 1000).

Before we had:

```java
interface RedBlackTree
    ^
   / \
Empty Node
```

I will come up with a suggestion later today how to simplify it.